dynosaur_derive/src/expand.rs

```diff
@@ -153,40 +140,77 @@ fn expand_fn_input(item_trait_generics: &Generics, trait_item_fn: &mut TraitItem
     }
 }
 
-pub(crate) fn expand_fn_output(trait_item_fn: &mut TraitItemFn) {
-    let sig = &mut trait_item_fn.sig;
+pub(crate) fn expand_sig_ret_ty_to_pin_box(sig: &mut Signature) {
+    let (arrow, ret) = expand_arrow_ret_ty(sig);
+    if let Some(asyncness) = sig.asyncness.take() {
+        sig.fn_token.span = asyncness.span;
+    }
+    sig.output = parse_quote! { #arrow ::core::pin::Pin<Box<dyn #ret + 'dynosaur>> };
+}
 
-    let (ret_arrow, ret) = expand_ret_ty(sig);
-    trait_item_fn.sig.output =
-        parse_quote! { #ret_arrow ::core::pin::Pin<Box<dyn #ret + 'dynosaur>> };
+pub(crate) fn expand_sig_ret_ty_to_rpit(sig: &mut Signature) {
+    let (arrow, ret) = expand_arrow_ret_ty(sig);
+    if let Some(asyncness) = sig.asyncness.take() {
+        sig.fn_token.span = asyncness.span;
+    }
+    sig.output = parse_quote! { #arrow impl #ret };
+}
+
+pub(crate) fn expand_ret_ty(sig: &Signature) -> TokenStream {
+    expand_arrow_ret_ty(sig).1
 }
 
-pub(crate) fn expand_ret_ty(sig: &Signature) -> (RArrow, TokenStream) {
-    let is_async = sig.asyncness.is_some();
-
-    if is_async {
-        let (ret_arrow, ret) = match &sig.output {
-            ReturnType::Default => (Token![->](Span::call_site()), quote!(())),
-            ReturnType::Type(arrow, ret) => (*arrow, quote!(#ret)),
-        };
-
-        (ret_arrow, quote! { ::core::future::Future<Output = #ret> })
-    } else {
-        if let ReturnType::Type(ret_arrow, ret) = &sig.output {
-            match &**ret {
-                Type::ImplTrait(TypeImplTrait { bounds, .. }) => (*ret_arrow, quote!(#bounds)),
-                _ => (
-                    Token![->](Span::call_site()),
-                    Error::new_spanned(&sig.output, "wrong return type").to_compile_error(),
-                ),
+pub(crate) fn expand_invoke_args(sig: &Signature, ufc: bool) -> Vec<TokenStream> {
+    let mut args = Vec::new();
+
+    for arg in &sig.inputs {
+        match arg {
+            FnArg::Receiver(_) => {
+                if !ufc {
+                    // Do not need & or &mut as this is at calling site
+                    args.push(quote! { self });
+                }
             }
-        } else {
-            (
+            FnArg::Typed(PatType { pat, .. }) => match &**pat {
+                Pat::Ident(arg) => {
+                    args.push(quote! { #arg });
+                }
+                _ => {
+                    args.push(
+                        Error::new_spanned(pat, "patterns are not supported in arguments")
+                            .to_compile_error(),
+                    );
+                }
+            },
+        }
+    }
+
+    args
+}
+
+fn expand_arrow_ret_ty(sig: &Signature) -> (RArrow, TokenStream) {
```

Comment:
This function sometimes returns a type and sometimes a list of bounds in the TokenStream, so the way you use it changes based on the input; I think it should be factored differently.